File: 07-browser-automation/golang/qq.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

const (
	qqTokenURL = "https://bots.qq.com/app/getAppAccessToken"
	qqAPIBase  = "https://api.sgroup.qq.com"
	qqIntents  = 1 << 25 // GROUP_AND_C2C: private C2C + group @ messages
)

var reMention = regexp.MustCompile(`<@!\d+>`)

type qqReplyCtx struct {
	kind     string // "c2c" | "group"
	targetID string
	msgID    string
}

type qqAdapter struct {
	appID   string
	secret  string
	handler func(ACPMessage)

	replyMu  sync.RWMutex
	replyCtx map[string]qqReplyCtx

	tokenMu sync.Mutex
	token   string
	tokenExp time.Time
}

func newQQAdapter() *qqAdapter {
	return &qqAdapter{
		appID:    os.Getenv("QQ_APP_ID"),
		secret:   os.Getenv("QQ_CLIENT_SECRET"),
		replyCtx: make(map[string]qqReplyCtx),
	}
}

func (q *qqAdapter) name() string { return "qq" }

func (q *qqAdapter) onMessage(h func(ACPMessage)) { q.handler = h }

// ── OAuth2 token ──────────────────────────────────────────────────────────────

func (q *qqAdapter) getAccessToken() (string, error) {
	q.tokenMu.Lock()
	defer q.tokenMu.Unlock()

	if q.token != "" && time.Now().Before(q.tokenExp.Add(-60*time.Second)) {
		return q.token, nil
	}

	body, _ := json.Marshal(map[string]string{"appId": q.appID, "clientSecret": q.secret})
	resp, err := http.Post(qqTokenURL, "application/json", bytes.NewReader(body))
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var data struct {
		AccessToken string `json:"access_token"`
		ExpiresIn   string `json:"expires_in"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return "", err
	}
	expiresIn, _ := strconv.Atoi(data.ExpiresIn)
	q.token = data.AccessToken
	q.tokenExp = time.Now().Add(time.Duration(expiresIn) * time.Second)
	return q.token, nil
}

func (q *qqAdapter) getGatewayURL(token string) (string, error) {
	req, _ := http.NewRequest("GET", qqAPIBase+"/gateway", nil)
	req.Header.Set("Authorization", "QQBot "+token)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	var data struct {
		URL string `json:"url"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return "", err
	}
	return data.URL, nil
}

// ── Send message back to QQ ───────────────────────────────────────────────────

func (q *qqAdapter) postJSON(token, url string, payload any) {
	body, _ := json.Marshal(payload)
	req, _ := http.NewRequest("POST", url, bytes.NewReader(body))
	req.Header.Set("Authorization", "QQBot "+token)
	req.Header.Set("Content-Type", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		fmt.Fprintf(os.Stderr, "[qq] POST %s error: %v\n", url, err)
		return
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		b, _ := io.ReadAll(resp.Body)
		fmt.Fprintf(os.Stderr, "[qq] POST %s failed: %d %s\n", url, resp.StatusCode, b)
	}
}

func (q *qqAdapter) send(reply AgentReply) {
	if reply.Type != "reply" {
		return
	}

	q.replyMu.RLock()
	ctx, ok := q.replyCtx[reply.SessionID]
	q.replyMu.RUnlock()
	if !ok {
		return
	}

	q.replyMu.Lock()
	delete(q.replyCtx, reply.SessionID)
	q.replyMu.Unlock()

	go func() {
		token, err := q.getAccessToken()
		if err != nil {
			fmt.Fprintf(os.Stderr, "[qq] getAccessToken: %v\n", err)
			return
		}
		msgSeq := time.Now().UnixMilli() % 65536
		payload := map[string]any{
			"content":  reply.Content,
			"msg_type": 0,
			"msg_id":   ctx.msgID,
			"msg_seq":  msgSeq,
		}
		var url string
		if ctx.kind == "c2c" {
			url = qqAPIBase + "/v2/users/" + ctx.targetID + "/messages"
		} else {
			url = qqAPIBase + "/v2/groups/" + ctx.targetID + "/messages"
		}
		q.postJSON(token, url, payload)
	}()
}

// ── QQ Gateway WebSocket ──────────────────────────────────────────────────────

func (q *qqAdapter) start() error {
	if q.appID == "" || q.secret == "" {
		fmt.Fprintln(os.Stderr, "[qq] QQ_APP_ID/QQ_CLIENT_SECRET not set, skipping QQ channel")
		return nil
	}
	go func() {
		for {
			if err := q.tryConnect(); err != nil {
				fmt.Fprintf(os.Stderr, "[qq] disconnected: %v — retry in 5s\n", err)
				time.Sleep(5 * time.Second)
			}
		}
	}()
	return nil
}

func (q *qqAdapter) tryConnect() error {
	token, err := q.getAccessToken()
	if err != nil {
		return fmt.Errorf("getAccessToken: %w", err)
	}
	gatewayURL, err := q.getGatewayURL(token)
	if err != nil {
		return fmt.Errorf("getGatewayURL: %w", err)
	}

	fmt.Fprintf(os.Stderr, "[qq] connecting to %s\n", gatewayURL)
	conn, _, err := websocket.DefaultDialer.Dial(gatewayURL, nil)
	if err != nil {
		return fmt.Errorf("dial: %w", err)
	}
	defer conn.Close()
	fmt.Fprintln(os.Stderr, "[qq] WebSocket connected")

	// Serialize concurrent writes (heartbeat goroutine + main loop both write).
	var writeMu sync.Mutex
	write := func(v any) error {
		writeMu.Lock()
		defer writeMu.Unlock()
		return conn.WriteJSON(v)
	}

	stopHB := make(chan struct{})
	defer close(stopHB)

	var lastSeq *int64

	for {
		_, raw, err := conn.ReadMessage()
		if err != nil {
			return err
		}

		var payload struct {
			Op int             `json:"op"`
			T  string          `json:"t"`
			D  json.RawMessage `json:"d"`
			S  *int64          `json:"s"`
		}
		if json.Unmarshal(raw, &payload) != nil {
			continue
		}
		if payload.S != nil {
			lastSeq = payload.S
		}

		switch payload.Op {
		case 10: // HELLO — start heartbeat then IDENTIFY
			var hello struct {
				HeartbeatInterval int `json:"heartbeat_interval"`
			}
			json.Unmarshal(payload.D, &hello)

			interval := time.Duration(hello.HeartbeatInterval) * time.Millisecond
			go func() {
				ticker := time.NewTicker(interval)
				defer ticker.Stop()
				for {
					select {
					case <-stopHB:
						return
					case <-ticker.C:
						write(map[string]any{"op": 1, "d": lastSeq}) //nolint:errcheck
					}
				}
			}()

			write(map[string]any{ //nolint:errcheck
				"op": 2,
				"d": map[string]any{
					"token":   "QQBot " + token,
					"intents": qqIntents,
					"shard":   []int{0, 1},
				},
			})

		case 0: // DISPATCH
			go q.handleEvent(payload.T, payload.D)

		case 9: // Invalid session
			fmt.Fprintln(os.Stderr, "[qq] invalid session, reconnecting")
			return fmt.Errorf("invalid session")

		case 7: // Reconnect
			fmt.Fprintln(os.Stderr, "[qq] server requested reconnect")
			return fmt.Errorf("server reconnect request")
		}
	}
}

func (q *qqAdapter) handleEvent(t string, d json.RawMessage) {
	switch t {
	case "C2C_MESSAGE_CREATE":
		var msg struct {
			Author struct {
				UserOpenid string `json:"user_openid"`
				ID         string `json:"id"`
			} `json:"author"`
			Content string `json:"content"`
			ID      string `json:"id"`
		}
		if json.Unmarshal(d, &msg) != nil {
			return
		}
		openid := msg.Author.UserOpenid
		if openid == "" {
			openid = msg.Author.ID
		}
		text := strings.TrimSpace(reMention.ReplaceAllString(msg.Content, ""))
		if text == "" {
			return
		}

		sessionID := "qq-c2c-" + openid
		q.replyMu.Lock()
		q.replyCtx[sessionID] = qqReplyCtx{kind: "c2c", targetID: openid, msgID: msg.ID}
		q.replyMu.Unlock()

		fmt.Fprintf(os.Stderr, "[qq] c2c from %s: %s\n", openid, text)
		q.handler(ACPMessage{
			ID:        newUUID(),
			SessionID: sessionID,
			Channel:   "qq",
			Content:   text,
			Timestamp: time.Now().UnixMilli(),
		})

	case "GROUP_AT_MESSAGE_CREATE":
		var msg struct {
			Author struct {
				MemberOpenid string `json:"member_openid"`
				ID           string `json:"id"`
			} `json:"author"`
			GroupOpenid string `json:"group_openid"`
			Content     string `json:"content"`
			ID          string `json:"id"`
		}
		if json.Unmarshal(d, &msg) != nil {
			return
		}
		openid := msg.Author.MemberOpenid
		if openid == "" {
			openid = msg.Author.ID
		}
		text := strings.TrimSpace(reMention.ReplaceAllString(msg.Content, ""))
		if text == "" {
			return
		}

		sessionID := "qq-group-" + msg.GroupOpenid
		q.replyMu.Lock()
		q.replyCtx[sessionID] = qqReplyCtx{kind: "group", targetID: msg.GroupOpenid, msgID: msg.ID}
		q.replyMu.Unlock()

		fmt.Fprintf(os.Stderr, "[qq] group %s from %s: %s\n", msg.GroupOpenid, openid, text)
		q.handler(ACPMessage{
			ID:        newUUID(),
			SessionID: sessionID,
			Channel:   "qq",
			Content:   text,
			Timestamp: time.Now().UnixMilli(),
		})
	}
}

File: 11-chronos/nodejs/src/channels/qq.ts
import WebSocket from 'ws';
import type { ChannelAdapter } from './types.ts';
import type { ACPMessage, AgentReply } from '../gateway/types.ts';
import { log } from '../logger.ts';

const TOKEN_URL = 'https://bots.qq.com/app/getAppAccessToken';
const API_BASE  = 'https://api.sgroup.qq.com';

// GROUP_AND_C2C (1<<25) intent — covers private C2C messages and group @ messages
const INTENTS = (1 << 25);

// ── Token 缓存 ────────────────────────────────────────────────────────────────

let tokenCache: { token: string; expiresAt: number } | null = null;

async function getAccessToken(appId: string, secret: string): Promise<string> {
  if (tokenCache && Date.now() < tokenCache.expiresAt - 60_000) {
    return tokenCache.token;
  }
  const res = await fetch(TOKEN_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ appId, clientSecret: secret }),
  });
  const data = await res.json() as { access_token: string; expires_in: number };
  tokenCache = { token: data.access_token, expiresAt: Date.now() + data.expires_in * 1000 };
  return data.access_token;
}

async function getGatewayUrl(accessToken: string): Promise<string> {
  const res = await fetch(`${API_BASE}/gateway`, {
    headers: { Authorization: `QQBot ${accessToken}` },
  });
  const data = await res.json() as { url: string };
  return data.url;
}

// ── 发送消息 ──────────────────────────────────────────────────────────────────

async function sendC2C(token: string, openid: string, content: string, msgId: string): Promise<void> {
  const res = await fetch(`${API_BASE}/v2/users/${openid}/messages`, {
    method: 'POST',
    headers: { Authorization: `QQBot ${token}`, 'Content-Type': 'application/json' },
    body: JSON.stringify({ content, msg_type: 0, msg_id: msgId, msg_seq: Date.now() % 65536 }),
  });
  if (!res.ok) console.error(`[qq] sendC2C failed: ${res.status} ${await res.text()}`);
}

async function sendGroup(token: string, groupOpenid: string, content: string, msgId: string): Promise<void> {
  const res = await fetch(`${API_BASE}/v2/groups/${groupOpenid}/messages`, {
    method: 'POST',
    headers: { Authorization: `QQBot ${token}`, 'Content-Type': 'application/json' },
    body: JSON.stringify({ content, msg_type: 0, msg_id: msgId, msg_seq: Date.now() % 65536 }),
  });
  if (!res.ok) console.error(`[qq] sendGroup failed: ${res.status} ${await res.text()}`);
}

// ── QQ Gateway WebSocket ──────────────────────────────────────────────────────

function connectGateway(
  appId: string,
  secret: string,
  onEvent: (t: string, d: unknown) => void,
): void {
  let heartbeatTimer: ReturnType<typeof setInterval> | null = null;
  let lastSeq: number | null = null;

  const cleanup = (ws: WebSocket) => {
    if (heartbeatTimer) { clearInterval(heartbeatTimer); heartbeatTimer = null; }
    if (ws.readyState === WebSocket.OPEN) ws.close();
  };

  const connect = async () => {
    try {
      const accessToken = await getAccessToken(appId, secret);
      const gatewayUrl = await getGatewayUrl(accessToken);
      if (!gatewayUrl) {
        console.error('[qq] failed to get gateway URL (check QQ_APP_ID / QQ_CLIENT_SECRET), giving up');
        return;
      }
      log(`[qq] connecting to ${gatewayUrl}`);

      const ws = new WebSocket(gatewayUrl);

      ws.on('message', (raw) => {
        let payload: { op: number; t?: string; d?: unknown; s?: number };
        try { payload = JSON.parse(raw.toString()); } catch { return; }
        const { op, t, d, s } = payload;
        if (s != null) lastSeq = s;

        switch (op) {
          case 10: { // HELLO
            const interval = (d as { heartbeat_interval: number }).heartbeat_interval;
            if (heartbeatTimer) clearInterval(heartbeatTimer);
            heartbeatTimer = setInterval(() => {
              if (ws.readyState === WebSocket.OPEN) {
                ws.send(JSON.stringify({ op: 1, d: lastSeq }));
              }
            }, interval);

            // IDENTIFY
            ws.send(JSON.stringify({
              op: 2,
              d: { token: `QQBot ${accessToken}`, intents: INTENTS, shard: [0, 1] },
            }));
            break;
          }
          case 0: // DISPATCH
            if (t) onEvent(t, d);
            break;
          case 11: // Heartbeat ACK
            break;
          case 9: // Invalid session
            console.error('[qq] invalid session, reconnecting…');
            cleanup(ws);
            setTimeout(connect, 5000);
            break;
          case 7: // Reconnect
            log('[qq] server requested reconnect');
            cleanup(ws);
            setTimeout(connect, 1000);
            break;
        }
      });

      ws.on('open', () => log('[qq] WebSocket connected'));
      ws.on('error', (err) => console.error('[qq] WebSocket error:', err));
      ws.on('close', (code) => {
        log(`[qq] WebSocket closed (code=${code}), reconnecting in 5s…`);
        cleanup(ws);
        setTimeout(connect, 5000);
      });
    } catch (err) {
      console.error('[qq] connect error:', err, '— retrying in 10s');
      setTimeout(connect, 10_000);
    }
  };

  connect();
}

// ── QQ Adapter ────────────────────────────────────────────────────────────────

export function createQQAdapter(): ChannelAdapter {
  const appId  = process.env['QQ_APP_ID'] ?? '';
  const secret = process.env['QQ_CLIENT_SECRET'] ?? '';

  let messageHandler: ((msg: ACPMessage) => void) | null = null;

  // 每条入站消息的回复上下文：sessionId → { type, targetId, msgId }
  const replyCtx = new Map<string, { type: 'c2c' | 'group'; targetId: string; msgId: string }>();

  return {
    name: 'qq',

    onMessage(h) { messageHandler = h; },

    send(reply: AgentReply): void {
      // delta 不发给 QQ（不支持流式）；只在最终 reply 时发送
      if (reply.type !== 'reply') return;
      const ctx = replyCtx.get(reply.sessionId);
      if (!ctx) return;
      replyCtx.delete(reply.sessionId);

      getAccessToken(appId, secret)
        .then((token) => ctx.type === 'c2c'
          ? sendC2C(token, ctx.targetId, reply.content, ctx.msgId)
          : sendGroup(token, ctx.targetId, reply.content, ctx.msgId))
        .catch((err) => console.error('[qq] send error:', err));
    },

    async start(): Promise<void> {
      if (!appId || !secret) {
        log('[qq] QQ_APP_ID / QQ_CLIENT_SECRET 未配置，跳过 QQ 渠道');
        return;
      }

      connectGateway(appId, secret, (t, d: any) => {
        if (t === 'C2C_MESSAGE_CREATE') {
          const msg = d.msg ?? d;
          const openid  = msg.author?.user_openid ?? msg.author?.id;
          const text    = (msg.content ?? '').replace(/<@!\d+>/g, '').trim();
          const msgId   = msg.id;
          if (!text) return;

          const sessionId = `qq-c2c-${openid}`;
          replyCtx.set(sessionId, { type: 'c2c', targetId: openid, msgId });
          log(`[qq] c2c from ${openid}: ${text}`);
          messageHandler?.({ id: crypto.randomUUID(), sessionId, channel: 'qq', content: text, timestamp: Date.now() });
        }

        if (t === 'GROUP_AT_MESSAGE_CREATE') {
          const msg         = d.msg ?? d;
          const groupOpenid = msg.group_openid;
          const openid      = msg.author?.member_openid ?? msg.author?.id;
          const text        = (msg.content ?? '').replace(/<@!\d+>/g, '').trim();
          const msgId       = msg.id;
          if (!text) return;

          const sessionId = `qq-group-${groupOpenid}`;
          replyCtx.set(sessionId, { type: 'group', targetId: groupOpenid, msgId });
          log(`[qq] group ${groupOpenid} from ${openid}: ${text}`);
          messageHandler?.({ id: crypto.randomUUID(), sessionId, channel: 'qq', content: text, timestamp: Date.now() });
        }
      });
    },
  };
}
